enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,910 @@
+type ServiceAliasConfig struct {
+    ...
+	// For route based setting of http response headers
+	SetHTTPResponseHeaders              []SetHTTPHeader
+	DeleteHTTPResponseHeaders           []DeleteHTTPHeader
+	// For route based setting of http request headers
+	SetHTTPRequestHeaders              []SetHTTPHeader
+	DeleteHTTPRequestHeaders           []DeleteHTTPHeader
+}
+```
+```go
+// SetHTTPHeader specifies an HTTP header that should be set.
+type SetHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+
+	// Value specifies the header value.
+	Value string
+
+	// Action specifies the action to be performed.
+	Action string
+}
+
+type DeleteHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+}
+```
+
+In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside router. So, we will run a loop
+on either `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` which will have the desired
+sequence. If request and response headers are provided in the API then first we run loop through request headers and then response headers.
+
+C) library-go code will be having validations for checking Route API fields.
+   We will be creating an immutable/hard-coded deny-list of HTTP response headers in [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141)
+   which won't be allowed to be set or deleted. For the initial enhancement, the HSTS `strict-transport-security` header name is included in this deny-list.
+   Only 64 items in request and response header list will be allowed.
+   Minimum length of the header value would have to be greater than 0 and less than 4096.
+   Sample fetchers for response dynamic header value allowed are "res.hdr" and "ssl_c_der".
+   Sample fetchers for request dynamic header value allowed are "req.hdr" and "ssl_c_der".
+   Converters allowed for request and response are "lower" and "base64".
+   Actions like Set/Delete for headers can't be set for passthrough routes.
+
+D) route/v1 of openshift/api will contain the validations using kubebuilder tags and CEL.
+The validations will be same as mentioned in point C).
+
+E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
+The validations will be same as mentioned in point C) except the one for passthrough routes.
+
+F) To validate the header value it has to match a regex.
+
+For the IngressController CR and Route CR (which is used for MicroShift), this validation is defined on the request and response action lists in openshift/api.
+For the Route API as handled by openshift-apiserver (which is used for OpenShift), this is defined in library-go.
```

Comment:
@miheer @JoelSpeed @Miciah I still don't quite understand why the regexes need to be declared in two different places.  This seems like something that we would want to avoid in the design, and if we purposely chose to do this, a reason why it couldn't be done in a more design-first manner?

Reply:
@candita this epic can't do anything about this. We are going by the design which exists. The openshift-apiserver does not use kubebuilder tags for validations for route so we need library-go to get these validations done.

Reply:
@candita just started a thread here https://github.com/openshift/library-go/pull/1449#discussion_r1191208997